crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,693 @@
+                        db.file_system().exists(&fs_path.join("__init__.pyi"))
+                    }
+                    TypeshedVersionsQueryResult::DoesNotExist => false,
+                }
+            }
+            (path, root) => unreachable!(
+                "The search path should always be the same variant as `self` (got: {path:?}, {root:?})"
+            )
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn parent(&self) -> Option<Self> {
+        Some(match self {
+            Self::Extra(ExtraPath(path)) => Self::extra_unchecked(path.parent()?),
+            Self::FirstParty(FirstPartyPath(path)) => Self::first_party_unchecked(path.parent()?),
+            Self::StandardLibrary(StandardLibraryPath(path)) => {
+                Self::standard_library_unchecked(path.parent()?)
+            }
+            Self::SitePackages(SitePackagesPath(path)) => {
+                Self::site_packages_unchecked(path.parent()?)
+            }
+        })
+    }
+
+    #[must_use]
+    fn ends_with_dunder_init(&self) -> bool {
+        match self {
+            Self::Extra(ExtraPath(path))
+            | Self::FirstParty(FirstPartyPath(path))
+            | Self::SitePackages(SitePackagesPath(path)) => {
+                path.ends_with("__init__.py") || path.ends_with("__init__.pyi")
+            }
+            Self::StandardLibrary(StandardLibraryPath(path)) => path.ends_with("__init__.pyi"),
+        }
+    }
+
+    #[must_use]
+    fn sans_dunder_init(self) -> Self {
+        if self.ends_with_dunder_init() {
+            self.parent().unwrap_or_else(|| match self {
+                Self::Extra(_) => Self::extra_unchecked(""),
+                Self::FirstParty(_) => Self::first_party_unchecked(""),
+                Self::StandardLibrary(_) => Self::standard_library_unchecked(""),
+                Self::SitePackages(_) => Self::site_packages_unchecked(""),
+            })
+        } else {
+            self
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn as_module_name(&self) -> Option<ModuleName> {
+        let mut parts_iter = match self.sans_dunder_init() {
+            Self::Extra(ExtraPath(path)) => ModulePartIterator::from_fs_path(path),
+            Self::FirstParty(FirstPartyPath(path)) => ModulePartIterator::from_fs_path(path),
+            Self::StandardLibrary(StandardLibraryPath(path)) => {
+                ModulePartIterator::from_fs_path(path)
+            }
+            Self::SitePackages(SitePackagesPath(path)) => ModulePartIterator::from_fs_path(path),
+        };
+        let first_part = parts_iter.next()?;
+        if let Some(second_part) = parts_iter.next() {
+            let mut name = format!("{first_part}.{second_part}");
+            for part in parts_iter {
+                name.push('.');
+                name.push_str(part);
+            }
+            ModuleName::new(&name)
+        } else {
+            ModuleName::new(first_part)
+        }
```

Comment:
Nit: I think we could move this to `ModuleName::from_components` where the argument is `impl IntoIterator<Item=&str>`